File: ocp-python/src/ocp_agent/http_client.py
"""
HTTP Client Enhancement - Make HTTP requests OCP-enabled.

Provides a simple HTTP client that automatically adds OCP context
headers to all requests.
"""

import functools
from typing import Dict, Any, Optional, Union, Callable
from urllib.parse import urlparse

from .context import AgentContext
from .headers import create_ocp_headers, extract_context_from_response


class OCPHTTPClient:
    """
    OCP-enabled HTTP client.
    
    Automatically adds OCP context headers to all requests and tracks
    interactions in the agent context.
    """
    
    def __init__(
        self, 
        context: AgentContext,
        auto_update_context: bool = True
    ):
        """
        Initialize OCP HTTP client.
        
        Args:
            context: Agent context to include in requests
            auto_update_context: Whether to automatically update context with interactions
        """
        self.context = context
        self.auto_update_context = auto_update_context
        
        # Use requests as our HTTP client
        import requests
        self.http_client = requests.Session()
    
    def _prepare_headers(self, headers: Optional[Dict[str, str]] = None) -> Dict[str, str]:
        """Prepare headers with OCP context."""
        ocp_headers = create_ocp_headers(self.context)
        
        if headers:
            # Merge with existing headers
            merged = headers.copy()
            merged.update(ocp_headers)
            return merged
        
        return ocp_headers
    
    def _log_interaction(self, method: str, url: str, response: Any = None) -> None:
        """Log the API interaction in context."""
        if not self.auto_update_context:
            return
        
        # Parse API endpoint
        parsed = urlparse(url)
        endpoint = f"{method.upper()} {parsed.path}"
        
        # Get response status if available
        result = None
        if response and hasattr(response, 'status_code'):
            result = f"HTTP {response.status_code}"
        elif response and hasattr(response, 'status'):
            result = f"HTTP {response.status}"
        
        # Add to context history
        self.context.add_interaction(
            action=f"api_call_{method.lower()}",
            api_endpoint=endpoint,
            result=result,
            metadata={"url": url, "domain": parsed.netloc}
        )
    
    def request(self, method: str, url: str, **kwargs) -> Any:
        """Make an HTTP request with OCP context."""
        # Prepare headers
        headers = kwargs.get('headers', {})
        kwargs['headers'] = self._prepare_headers(headers)
        
        # Make request
        response = self.http_client.request(method, url, **kwargs)
        
        # Log interaction
        self._log_interaction(method, url, response)
        
        return response
    
    def get(self, url: str, **kwargs) -> Any:
        """Make a GET request with OCP context."""
        return self.request('GET', url, **kwargs)
    
    def post(self, url: str, **kwargs) -> Any:
        """Make a POST request with OCP context."""
        return self.request('POST', url, **kwargs)
    
    def put(self, url: str, **kwargs) -> Any:
        """Make a PUT request with OCP context."""
        return self.request('PUT', url, **kwargs)
    
    def delete(self, url: str, **kwargs) -> Any:
        """Make a DELETE request with OCP context."""
        return self.request('DELETE', url, **kwargs)
    
    def patch(self, url: str, **kwargs) -> Any:
        """Make a PATCH request with OCP context.""" 
        return self.request('PATCH', url, **kwargs)


def wrap_api(
    base_url: str, 
    context: AgentContext,
    headers: Optional[Dict[str, str]] = None
) -> OCPHTTPClient:
    """
    Create an OCP-enabled client for a specific API.
    
    Args:
        base_url: Base URL for the API (e.g., "https://api.github.com")
        context: Agent context to include in requests
        headers: Additional headers to include in all requests
        
    Returns:
        OCP-enabled API client
        
    Example:
        >>> from ocp_agent import AgentContext, wrap_api
        >>> 
        >>> context = AgentContext(agent_type="debug_assistant")
        >>> github = wrap_api(
        ...     "https://api.github.com",
        ...     context,
        ...     headers={"Authorization": "token ghp_your_token"}
        ... )
        >>> issues = github.get("/search/issues", params={"q": "bug"})
    """
    # Set up base headers
    base_headers = headers.copy() if headers else {}
    
    # Create HTTP client with base URL
    import requests
    session = requests.Session()
    session.headers.update(base_headers)
    
    # Create OCP client
    ocp_client = OCPHTTPClient(context)
    ocp_client.http_client = session
    
    # Store base URL for convenience
    ocp_client.base_url = base_url.rstrip('/')
    
    # Override methods to handle relative URLs
    original_request = ocp_client.request
    
    def request_with_base_url(method: str, url: str, **kwargs):
        if not url.startswith('http'):
            url = f"{ocp_client.base_url}{url}"
        return original_request(method, url, **kwargs)
    
    ocp_client.request = request_with_base_url
    
    return ocp_client

File: ocp-javascript/src/http_client.ts
/**
 * OCP HTTP Client
 * 
 * HTTP client wrapper that automatically injects OCP context headers.
 */

import { AgentContext } from './context.js';
import { createOCPHeaders } from './headers.js';

/**
 * HTTP request options
 */
interface RequestOptions {
    params?: Record<string, string | number | boolean>;
    json?: any;
    headers?: Record<string, string>;
    timeout?: number;
}

/**
 * HTTP response wrapper
 */
export interface OCPResponse {
    status: number;
    statusText: string;
    ok: boolean;
    headers: Headers;
    data: any;
    text: string;
    json: () => Promise<any>;
}

/**
 * OCP HTTP Client
 * 
 * Wraps HTTP requests to automatically inject OCP context headers
 * and log interactions.
 */
export class OCPHTTPClient {
    private context: AgentContext;
    private autoUpdateContext: boolean;

    constructor(context: AgentContext, autoUpdateContext: boolean = true) {
        this.context = context;
        this.autoUpdateContext = autoUpdateContext;
    }

    /**
     * Prepare headers with OCP context.
     */
    private _prepareHeaders(additionalHeaders?: Record<string, string>): Record<string, string> {
        const ocpHeaders = createOCPHeaders(this.context);
        
        return {
            ...ocpHeaders,
            ...additionalHeaders
        };
    }

    /**
     * Log API interaction to context.
     */
    private _logInteraction(method: string, url: string, statusCode?: number, error?: Error): void {
        if (!this.autoUpdateContext) {
            return;
        }

        this.context.addInteraction(
            `api_call_${method.toLowerCase()}`,
            url,
            error ? `Error: ${error.message}` : (statusCode !== undefined ? `${statusCode}` : undefined),
            {
                method,
                status_code: statusCode,
                success: !error && statusCode ? statusCode >= 200 && statusCode < 300 : false,
                error: error ? error.message : undefined
            }
        );
    }

    /**
     * Make HTTP request.
     */
    async request(method: string, url: string, options: RequestOptions = {}): Promise<OCPResponse> {
        const { params, json, headers, timeout = 30000 } = options;

        // Build URL with query parameters
        const finalUrl = new URL(url);
        if (params) {
            for (const [key, value] of Object.entries(params)) {
                finalUrl.searchParams.append(key, String(value));
            }
        }

        // Prepare headers
        const finalHeaders = this._prepareHeaders(headers);
        
        // Add content-type for JSON
        if (json !== undefined) {
            finalHeaders['Content-Type'] = 'application/json';
        }

        // Create abort controller for timeout
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), timeout);

        try {
            const response = await fetch(finalUrl.toString(), {
                method: method.toUpperCase(),
                headers: finalHeaders,
                body: json !== undefined ? JSON.stringify(json) : undefined,
                signal: controller.signal
            });

            clearTimeout(timeoutId);

            // Read response text
            const text = await response.text();
            
            // Parse JSON if possible
            let data: any;
            try {
                data = JSON.parse(text);
            } catch {
                data = text;
            }

            const ocpResponse: OCPResponse = {
                status: response.status,
                statusText: response.statusText,
                ok: response.ok,
                headers: response.headers,
                data,
                text,
                json: async () => JSON.parse(text)
            };

            this._logInteraction(method, finalUrl.toString(), response.status);

            return ocpResponse;

        } catch (error) {
            clearTimeout(timeoutId);
            
            const err = error instanceof Error ? error : new Error(String(error));
            this._logInteraction(method, finalUrl.toString(), undefined, err);
            
            throw err;
        }
    }

    /**
     * GET request.
     */
    async get(url: string, options: RequestOptions = {}): Promise<OCPResponse> {
        return this.request('GET', url, options);
    }

    /**
     * POST request.
     */
    async post(url: string, options: RequestOptions = {}): Promise<OCPResponse> {
        return this.request('POST', url, options);
    }

    /**
     * PUT request.
     */
    async put(url: string, options: RequestOptions = {}): Promise<OCPResponse> {
        return this.request('PUT', url, options);
    }

    /**
     * DELETE request.
     */
    async delete(url: string, options: RequestOptions = {}): Promise<OCPResponse> {
        return this.request('DELETE', url, options);
    }

    /**
     * PATCH request.
     */
    async patch(url: string, options: RequestOptions = {}): Promise<OCPResponse> {
        return this.request('PATCH', url, options);
    }
}

/**
 * Create API-specific HTTP client.
 * 
 * @param context - Agent context
 * @param baseUrl - API base URL
 * @param headers - Optional headers to include in all requests
 * @returns Configured HTTP client
 */
export function wrapApi(context: AgentContext, baseUrl: string, headers?: Record<string, string>): OCPHTTPClient {
    const client = new OCPHTTPClient(context);
    
    // Normalize base URL (remove trailing slash)
    const normalizedBase = baseUrl.replace(/\/$/, '');
    
    // Helper to build full URL
    const buildUrl = (path: string): string => {
        // If path is absolute URL, return it unchanged
        if (path.startsWith('http://') || path.startsWith('https://')) {
            return path;
        }
        // Otherwise, concatenate with base URL
        return `${normalizedBase}${path}`;
    };
    
    // Override request method to handle base URL and additional headers (like Python)
    const originalRequest = client.request.bind(client);
    client.request = (method: string, path: string, options: RequestOptions = {}) => {
        const opts = { ...options };
        if (headers) {
            opts.headers = { ...headers, ...opts.headers };
        }
        return originalRequest(method, buildUrl(path), opts);
    };

    return client;
}
